### PR TITLE
Moving common AWS Connection settings into a re-usable tab

### PR DESCRIPTION
**Pull request by kiiadi**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description

Create consistent "AWS Connection" run configuration settings tab.

## Motivation and Context

At least two (and possibly more in the future) run configurations require (or allow) the customer to configure which credential/region the configuration is being run in - currently they're inconsistent.

## Screenshots (if appropriate)
Screen Shot 2019-09-05 at 5 51 39 AM
Screen Shot 2019-09-05 at 5 51 33 AM
Screen Shot 2019-09-05 at 5 51 16 AM

## Checklist

- [ ] I have read the **README** document
- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `gradlew check` succeeds
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/remote/RemoteLambdaRunSettingsEditor.kt [hunk 1]
```diff
@@ -8,37 +8,33 @@ import com.intellij.openapi.application.runInEdt
...
-    private fun updateFunctions() {
+    internal fun updateFunctions(region: AwsRegion?, credentialProvider: String?) {
+        val (oldRegion, oldCredentialProvider) = settings.getAndUpdate { (_, _) -> region to credentialProvider }
+        if (oldRegion == region && oldCredentialProvider == credentialProvider) return
         view.functionNames.selectedItem = null
         view.setFunctionNames(emptyList())
 
-        val credentialProviderId = view.credentialSelector.getSelectedCredentialsProvider() ?: return
-        val region = view.regionSelector.selectedRegion ?: return
+        credentialProvider ?: return
```
Comment (awschristou) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/remote/RemoteLambdaRunSettingsEditor.kt]: If credentialProvider or region are null, would you want to clear out `functionName` before returning?

Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/remote/RemoteLambdaRunSettingsEditor.kt]: ported to the `ResourceSelector` now we don't need any of this logic - it's all handled for us

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/LambdaRunConfigurationBase.kt [hunk 2]
```diff
@@ -13,35 +12,24 @@ import com.intellij.openapi.fileEditor.FileDocumentManager
...
     private val configFactory: ConfigurationFactory,
     id: String
-) : LocatableConfigurationBase<T>(project, configFactory, id),
+) : AwsConnectionsRunConfigurationBase<T>(project, configFactory, id),
     RunConfigurationWithSuppressedDefaultRunAction,
     RunConfigurationWithSuppressedDefaultDebugAction {
 
-    protected abstract val lambdaOptions: BaseLambdaOptions
-
-    final override fun readExternal(element: Element) {
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/LambdaRunConfigurationBase.kt]: Were the only settings getting persisted here the AWS Account settings?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/local/LocalLambdaRunConfiguration.kt [hunk 3]
```diff
@@ -54,16 +56,18 @@ class LocalLambdaRunConfigurationFactory(configuration: LambdaRunConfiguration)
 class LocalLambdaRunConfiguration(project: Project, factory: ConfigurationFactory) :
     LambdaRunConfigurationBase<LocalLambdaOptions>(project, factory, "SAM CLI"),
     RefactoringListenerProvider {
+
     companion object {
         private val logger = getLogger<LocalLambdaRunConfiguration>()
     }
 
-    override val lambdaOptions = LocalLambdaOptions()
+    override val serializableOptions = LocalLambdaOptions()
```
Comment (bryceitoc9) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/local/LocalLambdaRunConfiguration.kt]: Why was this name changed? It looks like none of the functionality changed, and the class name hasn't changed.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/remote/RemoteLambdaRunConfiguration.kt [hunk 4]
```diff
@@ -24,9 +28,15 @@ class RemoteLambdaRunConfigurationFactory(configuration: LambdaRunConfiguration)
 class RemoteLambdaRunConfiguration(project: Project, factory: ConfigurationFactory) :
     LambdaRunConfigurationBase<RemoteLambdaOptions>(project, factory, "Remote") {
 
-    override val lambdaOptions = RemoteLambdaOptions()
+    override val serializableOptions = RemoteLambdaOptions()
```
Comment (bryceitoc9) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/remote/RemoteLambdaRunConfiguration.kt]: Same as above, should we change the class name?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/local/LocalLambdaRunConfiguration.kt (hunk 3 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/local/LocalLambdaRunConfiguration.kt]: parent class isnt tied to lambda

Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/local/LocalLambdaRunConfiguration.kt]: it's overriding a val that's specified in a generic `AwsConnectionsRunConfigurationBase` class that is no longer lambda specific.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/LambdaRunConfigurationBase.kt (hunk 2 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/LambdaRunConfigurationBase.kt]: This persisted all the settings in the UI

Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/LambdaRunConfigurationBase.kt]: no - but this just uses bean-based serialization of the class - so any properties on the actual implementation should get persisted.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/remote/RemoteLambdaRunConfiguration.kt (hunk 4 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/remote/RemoteLambdaRunConfiguration.kt]: `RemoteLambdaOptions` is an implementation that extends `BaseAwsConnectionOptions`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/local/LocalLambdaRunConfiguration.kt (hunk 3 above)
Comment (bryceitoc9) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/local/LocalLambdaRunConfiguration.kt]: Nvm didn't see the "override"

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/LambdaRunConfigurationBase.kt (hunk 2 above)
Comment (hunterwerlla) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/LambdaRunConfigurationBase.kt]: Right so it seems weird to put this in the AWS Settings tab: we are persisting all settings through the AWS Settings tab?

**Pull request merged**